[MIGRATED TO PR #693] Japanese Layout in Menus (Swaps Accept/Confirm button)

AL2009man opened the pull request:
note: this PR requires https://github.com/fgsfdsfgs/perfect_dark/pull/627

What `SDL_HINT_GAMECONTROLLER_USE_BUTTON_LABELS` does is that SDL's Controller API will follow on Button Labels, by default: it is set to `1`. Consequences: the modern Nintendo Switch layout will be using the Nintendo Layout across Menu Navigation and Gameplay. 

> ```
>      (Y)              (X)
>   (X)   (B)        (Y)   (A)
>      (A)              (B)
> 
>   Standard         Nintendo
> ```

Beyond copy-pasting: it results in muscle memory. The initial workaround would be to rebind to the Key Binding section, but I argue it ain't very nice. Taking a page from https://github.com/yquake2/yquake2/pull/1183/commits/9e07dc2485a68ebe1a65e90d66d127a9c4f880e0, I took a stab at it myself. the results is very mixed and feels a bit hacky-- but it should be easier on the UI/UX front. 

do note that it only applies to both "ACCEPT_BUTTON/CK_ACCEPT" and "CANCEL_BUTTON/CK_CANCEL" action at the moment, no plans to touch `Use / Accept/Cancel` action

While both toggles aren't accessible in-game, both `JapaneseButtonLayout` (will only swap Menu Accept/Cancel) is always accessible within `pd.ini`.

~~As of this writing: the only holdout is the deadzone itself. Right now: both `DEFAULT_DEADZONE` (4096) and `DEFAULT_DEADZONE_RY` (6144) controls baseline deadzone slider, but if using a legitament Nintendo Switch controller (not the third-party kind): both left and right stick will drift. That part might be easily addressable by increasing both sliders, or add a NintendoSwitch-specific deadzone slider to remedy the problem.~~

AL2009man committed:
added gitignore

AL2009man committed:
fixed Button Layout for Nitntendo Controllers

`SDL_HINT_GAMECONTROLLER_USE_BUTTON_LABELS` will follow the Button Labels, which causes issues in games that uses it. This commit will turn it off by default, but the default can be restored via `useNintendoLayout` within pd.ini file

AL2009man committed:
fixed Face button layout for Nintendo Switch, Japanese face button layout

`SDL_HINT_GAMECONTROLLER_USE_BUTTON_LABELS` is set to 0, which will follow the positional diamond-shaped face buttons on the game controller. making everything more streamlined and easier to translate across all modern contorllers. This solves a specific issue with

Japanese-style face buttons for `BUTTON_UI_ACCEPT`/`BUTTON_UI_CANCEL` actions has been added. A bit of a hacky workaround, but it'll do for now. By default: it'll only kicks in when using a Nintendo Switch controller, but `JapaneseButtonLayout` is accessible within `pd.ini`

AL2009man committed:
remove "/vscode" from .gitignore

leftovers due to switching to Visual Studios code

AL2009man committed:
updated documentation for the new button layout

AL2009man committed:
changed title from `inputIsSwitchController` to `inputIsNintendoSwitchController`

AL2009man committed:
putting guardrails for GitHub Action CI

AL2009man committed:
added UI Menu text to reflect the Japanese layout

when going to the Keybind page, you'd see "[JPN LAYOUT - B/A]" on UI Accept/Cancel text, UI Text will change dynamically.

AL2009man committed:
another attempt to fix build-i686-linux

AL2009man commented:
note: this PR also contains issues with build-i686-linux. it won't compile properly, but based on initial look: I believe this is related to using older version of SDL for some odd reason. 

As this PR is more of less an foundation for adding proper Button Labels support, this PR can't be pushed until that's be solved.

fgsfdsfgs commented:
The button label swapping is a good feature, however I feel like the "Japanese layout" thing can be just as well accomplished by just swapping the bindings on the Accept/Cancel buttons. In any case, it would also be best for them to be in separate PRs.

The i686 problem is likely because the version of SDL2 available (or just used) for i686-linux is older and does not have this feature. It is in general best to guard these things with `SDL_VERSION` checks.

AL2009man commented:
> The A/B/X/Y swapping is a good feature, however I feel like the "Japanese layout" thing can be just as well accomplished by just swapping the bindings on the Accept/Cancel buttons in the Extended Controls menu. I don't really want to add to the already complicated and confusing input logic just to make it very slightly easier to do, as it's already a pretty niche thing. In any case, it would also be best for them to be in separate PRs.

Not a new feature, **a necessity**. 

After playing with my Nintendo Switch controller, **it follows the Nintendo layout in a literal way**. I could rebind them to correct it, but since I often *switch* (no pun intended) controller types: it retroactively changes Player 1/2/3/4 bindings in the process. This is not ideal. 

which is why `useJapaneseLayout` is a thing. What it really does is reversing "ACCEPT_BUTTON" / "CANCEL_BUTTON". Under the hood; it's really the same binding. This is only done for backwards compatibility.

I was originally going for the hardcoded approach, but based on feedback: I might revise it by opting for Button Input instead of a action.

> I won't add controller-specific deadzone settings, that's a bit too insane. It also depends on the specific Switch controller. I tested these settings against the joycons on my own early model Switch and it was fine. It's generally not possible to account for everything, which is why the deadzone is user-adjustable.

Which is why I'm deciding whenever or not:

* Increase overall minimum deadzone to better account for first party controllers that wasn't calibrated or requires dedicated hardware (which is my case with Switch Pro Controller)
* change default In-Game Deadzone to something a bit higher.

AL2009man committed:
replaced Action-based swaps with hardcoded button inputs when using Japanese layout

also changed the guard function for GitHubAction CI

AL2009man committed:
ditched hardcoded button inputs for the time being

reverted back to the game action-based setup for the time being. this might be revised when a separate PR changes the Button Input labels. (make it easier to troubleshoot

AL2009man committed:
updated default Joystick deadzone to 6500 / 6

to better accout for uncalibrated joysticks: the default joystick slider will be increasing to 6 percent. This should address a specific issue for first-party Nintendo Switch Pro Controller users who can't calibrate their joystick drifts.

AL2009man committed:
removed NintendoLayout for the time being

this will be moved to a separate PR (https://github.com/fgsfdsfgs/perfect_dark/pull/627).

AL2009man committed:
Restored original default deadzone to prepare for future SDL2 release

https://github.com/libsdl-org/SDL/pull/13260 addresses this particular issue for Nintendo Switch Pro Controller's center deadzone, this: the original default deadzone has been restored to 4 percent and 6 percent

AL2009man committed:
Adjustments Japanese Layout input handling

Cleaned up and fixed the remaining bugs within the input system. Options Menu's Bind UI should now correctly reflect the Japanese Layout

AL2009man committed:
attempt fix regression for i686-linux

AL2009man committed:
restored previous workaround for GitHub Action compiling

AL2009man committed:
replaced SDL Buttons with CK_BUTTON's

AL2009man committed:
removed Getter and Setter for Japanese Layout for now

temporarily removed Get and Set functions, as it's leftover code for a pending Menu UI toggle..

AL2009man committed:
Simplified Japanese Layout setup

To better prepare for `SDL_GetGamepadButtonLabel`/SDL3: the entire setup has been simplified by combining `inputIsJapaneseLayoutActive` and `inputRemapUIButton` into one, while giving it a new name

AL2009man committed:
attempt whitespace fix

AL2009man committed:
code cleanup

AL2009man commented:
> The A/B/X/Y swapping is a good feature, however I feel like the "Japanese layout" thing can be just as well accomplished by just swapping the bindings on the Accept/Cancel buttons in the Extended Controls menu. **I don't really want to add to the already complicated and confusing input logic just to make it very slightly easier to do, as it's already a pretty niche thing.** In any case, it would also be best for them to be in separate PRs.

ok, after experimented with `SDL_Get GamepadButtonLabel` (SDL3 only): it's actually a lot easier to avoid making a complicated/confusing input logic by simply making this as our basis on Button UI Swap.

Example code: 
```
        if (SDL_GetGamepadButtonLabel(gamepad, SDL_GAMEPAD_BUTTON_SOUTH) == SDL_GAMEPAD_BUTTON_LABEL_B) {
            flipped_buttons = true;
        } else {
            flipped_buttons = false;
        }
```

This gets rid of the need for Nintendo Switch-specific detection system as `SDL_GetGamepadButtonLabel(` covers that automatically. For that: I simplify the Action

AL2009man committed:
fixed `SDL_VERSIONATLEAST number

double-checked SDL2 version to addrsss incorrect build release`

AL2009man committed:
swapped `JapaneseButtonLayout` order

to make things a bit more consistent, the order on how to enable or disable has been swapped

AL2009man committed:
reorder `JapaneseButtonLayout` order, to properly use `-1` Auto

 this time: actually follow the order

AL2009man committed:
simplified ConfirmCancel function

this reduces the sheer amount of bloat, as most of the functions are moved to a helper function. This also fixes a bug where the Swap applies to all Actions while rebinding.
